### ISSUE TITLE
cross-compiling to aarch64-unknown-linux-gnu uses wrong linker

### ISSUE THREAD

**Issue by andrewdavidmackenzie** [axodotdev/cargo-dist#1378]
Hi, request for help.

In my release the build-local-artifacts step fails for two arm arch builds I have, when building ring, thus:

```
   Compiling ring v0.17.8
   Compiling parking v2.2.0
The following warnings were emitted during compilation:

warning: ring@0.17.8: Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-gnueabihf-gcc` installed?
warning: ring@0.17.8: Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-gnueabihf-gcc` installed?
```

Full job here: https://github.com/andrewdavidmackenzie/pigg/actions/runs/10597707922/job/29368504382

Here is my [Cargo.toml](https://github.com/andrewdavidmackenzie/pigg/blob/master/Cargo.toml)
Here is my [release.yml](https://github.com/andrewdavidmackenzie/pigg/blob/master/.github/workflows/release.yml)

How do I fix that?

**Comment by andrewdavidmackenzie** [comment]
Maybe comments in this thread are part of my solution?
https://github.com/axodotdev/cargo-dist/issues/74

"Add [gcc-aarch64-linux-gnu](https://packages.ubuntu.com/focal/gcc-aarch64-linux-gnu) to cargo-dist's dependencies.apt table. This installs a cross-compilation toolchain which can be invoked as aarch64-linux-gnu-gcc."

**Comment by andrewdavidmackenzie** [comment]
I tried that, and things progressed but still this failure I need to look at more:

https://github.com/andrewdavidmackenzie/pigg/actions/runs/10599058041/job/29373158574

**Comment by mistydemeo** [comment]
It looks like your `armv7-unknown-linux-gnueabihf` job is running into a bug/limitation in cargo-dist.

When we determine how to install dependencies, we look at the target triple. Since `armv7-unknown-linux-gnueabihf` isn't a target triple we recognize, we end up skipping package installation on that platform. We'll look at fixing this, and may have other fixes for your usecase in the future. In the meantime, our [customizing build setup feature](https://opensource.axo.dev/cargo-dist/book/ci/customizing.html#customizing-build-setup), which lets you run extra steps before your build begins, may be of use in making sure you can get the toolchain you need to target that platform.

Your aarch64-unknown-linux-gnu build looks like it's using the toolchain you installed, and is failing for reasons I'm not able to help with.

**Comment by andrewdavidmackenzie** [comment]
I tried to customize, but suspect it's not being used.

In the armv7 job [here](https://github.com/andrewdavidmackenzie/pigg/actions/runs/10612259267/job/29413546717) the "Install Dependencies" step is empty.

The aarch64 job [here](https://github.com/andrewdavidmackenzie/pigg/actions/runs/10612259267/job/29413546390) is doing the gcc install, but I suspect it's picking that up from my [meta-data in Cargo.toml](https://github.com/andrewdavidmackenzie/pigg/blob/master/Cargo.toml#L125)

Did I get the addition of the setup right?
- Cargo.toml addition [here](https://github.com/andrewdavidmackenzie/pigg/blob/master/Cargo.toml#L113)
- added gcc-install.yml file [here](https://github.com/andrewdavidmackenzie/pigg/blob/master/.github/workflows/gcc-install.yml)

**Comment by andrewdavidmackenzie** [comment]
Tried a bunch of things today... no luck.

It's trying to link using "cc" and failing. [job log](https://github.com/andrewdavidmackenzie/pigg/actions/runs/10619659739/job/29437740338)

I'm stuck, and will have to remove these arm targets from my cargo-dist list of targets I fear...

**Comment by andrewdavidmackenzie** [comment]
When searching for the error reported in my [build log](https://github.com/andrewdavidmackenzie/pigg/actions/runs/10619659739/job/29437740338):

I see posts such as [this](https://stackoverflow.com/questions/51476414/relocations-in-generic-elf-em-183) that say my error is due to

`"This is invoking the wrong linker (system linker), and most likely also invoking the wrong compiler ...." `

**Comment by andrewdavidmackenzie** [comment]
I found this bug which mentioned the same error:

https://github.com/rust-lang/rust/issues/82519

The error message [here](https://github.com/andrewdavidmackenzie/pigg/actions/runs/10657310588/job/29536993976) is:
```
 = note: /usr/bin/ld: /home/runner/work/pigg/pigg/target/aarch64-unknown-linux-gnu/dist/deps/piggui-cb0337339124d967.4g3dfa9548xp8fq1f7dediq7i.rcgu.o: Relocations in generic ELF (EM: 183)
```

certainly is from using "ld" and not an aarch64 linker - so seems to be the same issue.

NOTE: I have modified the title of this issue to reflect this one of my two cargo-dist release build errors related to arm, I have filed an issue with `ring` for the other.

**Comment by andrewdavidmackenzie** [comment]
No change. I'm out of ideas now.

**Comment by Gankra** [comment]
Apologies, Misty is on vacation this week.

Starting next week I'll be picking up 

* #74 

again (this time based on maturin, which some of our users are very-manually using). I think that's the only really practical way for us to support these cross-compiles.

**Comment by andrewdavidmackenzie** [comment]
Was wondering if work started on #74 and if we can help in any way?

**Comment by andrewdavidmackenzie** [comment]
I know that you are working on a totally new approach for cross compiling, but meanwhile I was looking at my failures again:

aarch64 build:

https://github.com/andrewdavidmackenzie/pigg/actions/runs/10901160279/job/30250372834

I can see that in the "Install dependencies" step it is using "apg-get" to install the aarch64 toolchain, presumably based on my meta.d.ist.dependencies.apt key in my Cargo.toml:

```
[workspace.metadata.dist.dependencies.apt]
gcc-aarch64-linux-gnu = { version = '*', targets = ["aarch64-unknown-linux-gnu"] }
gcc-arm-linux-gnueabihf = { version = '*', targets = ["armv7-unknown-linux-gnueabihf"] }
# Try this variation to see if it works
arm-linux-gnueabihf-gcc = { version = '*', targets = ["armv7-unknown-linux-gnueabihf"] }
```

But in the armv7 build:

https://github.com/andrewdavidmackenzie/pigg/actions/runs/10901160279/job/30250373154

I don't see it attempting to install either of my two attempts at an armv7 toolchain.

And that could be one reason that build fails with:

```
cargo:warning=Compiler family detection failed due to error: ToolNotFound: Failed to find tool. Is `arm-linux-gnueabihf-gcc` installed?
```

Do you know why it is not attempting to install the specified dependency for the "armv7-unknown-linux-gnueabihf" target?

**Comment by andrewdavidmackenzie** [comment]
I have been able to get the aarch64 build via cargo-dist to work.
For the armv7 build there is a problem with non-installation of the linker dependency.
I have opened a new issue for that here https://github.com/axodotdev/cargo-dist/issues/1417

So, I think we should close this one in favor of https://github.com/axodotdev/cargo-dist/issues/1417

If that issue could be fixed, our cross compiling needs would be met with the current implementation.

FYI @Gankra
